### ISSUE TITLE
Waiting for services to be ready

### ISSUE THREAD

**Issue by javabean** [joyent/containerpilot#196]
On on the main sticking points in orchestrating a datacenter is to correctly manage dependencies between services. This means not only creating them in the correct order, but also to be able to wait until a service is ready for action before starting a dependent service!

We are nearly there with ContainerPilot, but we still need 2 small RFEs to achieve nirvana:
* the ability to wait until all `backends` services are available (i.e. published in Consul / etcd / …) before starting `preStart`: new optional configuration key `startWaitsForBackends` (defaults to `false` for backward compatibility)
* the ability to delay publishing our own `services` until the main process is ready to serve.  
  This can be done in 2 ways:  
  * waiting until the first `health`check passes
  * creating a new `waitForServiceReady` lifecycle event, which exit will give the point in time where we can register our own services in Consul (and obviously delaying any `health` task until this time)

**Comment by tgross** [comment]
@javabean our position has been that this kind of behavior needs to be in the application rather than ContainerPilot. The reason for this is that all services need to be able to behave in a correct way even if their dependencies are gone (whatever "correct" means for your stack).

As an example of this you might look at the Nginx configuration template in [autopilotpattern/workshop](https://github.com/autopilotpattern/workshop/blob/master/nginx/nginx.conf.ctmpl). In that example, if one of the services is missing then requests to that service will receive 404 because the `if` block will eval to false and we won't write the location block in question. One of Joyent's customers is using this but has decided that 404 isn't the behavior they want in _their_ stack and so instead they have an `else` block that write a location block that returns 503. So if all their instances of a given service aren't available then they return 503 rather than 404 because that's what they've decided their application requires.

ContainerPilot is intentionally unopinionated on how your application needs to behave in these scenarios because we think that belongs in the hands of the application developer.

**Comment by javabean** [comment]
@tgross I fully agree with you, and I strive to keep a high external context leniency in my own applications.

Unfortunately, I also have to deal with 3rd-parties applications which are not, to put it mildly, very well behaved when e.g. the database is not readily available. For those, making sure only up-and-ready services are exposed in Consul (in contrast of "server is here, but service may come in a while") would be absolutely fantastic!

Maybe I'm being too greedy, wanting to use ContainerPilot everywhere? :-D

**Comment by tgross** [comment]
> Unfortunately, I also have to deal with 3rd-parties applications which are not, to put it mildly, very well behaved when e.g. the database is not readily available.

Sure, but by saying "the application" is responsible what I'm really saying is "the application and the user-defined behaviors the application developer has provided in the way of in-container handlers."

So in the specific use case you're considering here, this is what a `preStart` hook can be used for. Have it poll for its dependencies by hitting the service catalog (via `curl`, let's say) and then only exit the `preStart` when they're available.

**Comment by javabean** [comment]
This is clearer expressed like such, thanks. I can indeed use `preStart` to poll for required services readiness.
Subsidiary question: will `onChange` events fire while stuck in `preStart`? (AKAIK no; it would be good to merge those Joyent blog posts information & graphics into the main documentation :-))

That said, what happens when an application takes a long time to start up? I recon:
* `preStart`
* application starts starting up
* service published in Consul
* 1st (and following) `health`check fails -> service unregistered in Consul
* application finally ready for service
* next `health`check OK -> re-registers service in Consul

What would be your take to avoid this "publish + unpublish + republish" thing? That's 3 container reconfigurations for no good (since the service was never available anyway), for potentially dozens of dependent containers!

Final bonus question: does ContainerPilot de-register the service before calling `preStop`? I can't find any such thing in the [documentation](https://www.joyent.com/blog/applications-on-autopilot). Should I open a new issue for this?

**Comment by misterbisson** [comment]
I've opened https://github.com/joyent/containerpilot/issues/197 to amend the documentation to address the questions here.

**Comment by javabean** [comment]
Brilliant! I'll try to contribute a documentation PR in the coming weeks.

May I suggest that we systematically de-register services in case of main application crash? I know we're on a Docker-managed timer here, so this could be done concurrently to `preStop` / `postStop` in a separate thread.

**Comment by tgross** [comment]
I've opened a discussion of the proposal for trying to de-registering services on crash in https://github.com/joyent/containerpilot/issues/200. I'm going to close this issue in lieu of #197 and #200.
